programs/mango-v4/src/instructions/perp_place_order.rs

```diff
@@ -109,8 +110,22 @@ pub fn perp_place_order(ctx: Context<PerpPlaceOrder>, order: Order, limit: u8) -
 
     let now_ts: u64 = Clock::get()?.unix_timestamp.try_into().unwrap();
 
-    // TODO apply reduce_only flag to compute final base_lots, also process event queue
-    require!(order.reduce_only == false, MangoError::SomeError);
+    let effective_pos = account
+        .perp_position(perp_market_index)?
+        .effective_base_position_lots();
+    order.max_base_lots = if order.reduce_only || perp_market.is_reduce_only() {
+        if (order.side == Side::Bid && effective_pos > 0)
+            || (order.side == Side::Ask && effective_pos < 0)
+        {
+            0
```

Comment:
This is another place where we can debate error vs do-nothing-success. 

- If the market is reduce-only and the user sends a non-reduce-only order, wouldn't we expect an error here?
- If the user sends a reduce-only order, then silent success is what I'd expect.

A similar thing applies to the reduction of the order max_base_lots below, imo. If the user didn't request reduce-only, then I'd rather error instead of silently reducing their order amount.

Reply:
Done! Please re-review flow in place perp order ix

Reply:
Almost. I'm thinking of integrators that want to close their position and forget to send reduce-only. In reduce-only markets, their order would now be an error, even if it would work without adjustment.

I think for best behavior, we should accept orders that are not tagged reduce-only even when the market is reduce-only -- but only as long as they can go though unadjusted.

Reply:
Done